flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/expressions/RexNodeConverter.java

```diff
@@ -107,16 +310,16 @@ public RexNode visit(CallExpression call) {
 				.collect(Collectors.toList());
 	}
 
-	private RexNode translateScalarCall(FunctionDefinition def, List<Expression> children) {
+	private RexNode visitBuiltInFunc(FunctionDefinition def, List<Expression> children) {
 
-		if (def.equals(BuiltInFunctionDefinitions.CAST)) {
+		if (BuiltInFunctionDefinitions.CAST.equals(def)) {
 			RexNode child = children.get(0).accept(this);
 			TypeLiteralExpression type = (TypeLiteralExpression) children.get(1);
 			return relBuilder.getRexBuilder().makeAbstractCast(
 					typeFactory.createFieldTypeFromLogicalType(
 							type.getOutputDataType().getLogicalType().copy(child.getType().isNullable())),
 					child);
-		} else if (def.equals(BuiltInFunctionDefinitions.REINTERPRET_CAST)) {
+		} else if (BuiltInFunctionDefinitions.REINTERPRET_CAST.equals(def)) {
```

Comment:
Why not reference equality? The documentation on `BuiltinFunctionDefinition` says we should use reference equality.

Reply:
One problem is that if `BuiltInFunctionDefinition` is serialized and deserialized, its references will vary. Why does it not implement `readObject` and `writeObject` to keep real singleton?

Reply:
Oh, I got it, `BuiltInFunctionDefinition` is not Serializable.

Reply:
They should never be serialized. They are purely API facing classes that should not be shipped to the cluster.

Reply:
good catch.